sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -159,4 +180,70 @@ ShardRecordsIterator createShardIterator(SimplifiedKinesisClient kinesis,
     return new ShardRecordsIterator(checkpoint, kinesis);
   }
 
+  /**
+   * Waits until all records read from given shardRecordsIterator are taken from
+   * {@link #recordsQueue} and acked.
+   * Uses {@link #numberOfRecordsInAQueueByShard} map to track the amount of remaining events.
+   */
+  private void waitUntilAllShardRecordsRead(ShardRecordsIterator shardRecordsIterator)
+      throws InterruptedException {
+    while (!allShardRecordsRead(shardRecordsIterator)) {
+      Thread.sleep(TimeUnit.SECONDS.toMillis(1));
+    }
+  }
+
+  private boolean allShardRecordsRead(final ShardRecordsIterator shardRecordsIterator) {
+    return numberOfRecordsInAQueueByShard.get(shardRecordsIterator.getShardId()).get() == 0;
+  }
+
+  private void readFromSuccessiveShards(final ShardRecordsIterator closedShardIterator)
+      throws TransientKinesisException {
+    List<ShardRecordsIterator> successiveShardRecordIterators = closedShardIterator
+        .findSuccessiveShardRecordIterators();
+
+    if (!successiveShardRecordIterators.isEmpty()) {
+
+      // Handle one closedShardIterator at a time to keep shardIteratorsMap,
+      // recordsQueue and numberOfRecordsInAQueueByShard in sync.
+      synchronized (this) {
+        ImmutableMap.Builder<String, ShardRecordsIterator> shardsMap = ImmutableMap.builder();
+        Iterable<ShardRecordsIterator> allShards = Iterables
+            .concat(shardIteratorsMap.values(), successiveShardRecordIterators);
+        for (ShardRecordsIterator iterator : allShards) {
+          if (!closedShardIterator.getShardId().equals(iterator.getShardId())) {
+            shardsMap.put(iterator.getShardId(), iterator);
+          }
+        }
+        shardIteratorsMap = shardsMap.build();
+        numberOfRecordsInAQueueByShard.remove(closedShardIterator.getShardId());
+
+        BlockingQueue<KinesisRecord> newRecordsQueue = new ArrayBlockingQueue<>(
```

Comment:
Here are we simply draining the old contents of the queue into the new instance of the queue? Why can't we keep using the existing queue?

Reply:
The queue capacity is adjusted to the changing number of shards so it’s replaced with every change of shards number. It would be tricky to choose one queue capacity that would fit in all cases so I decided to implement such resizing. Also, using unbounded queue could cause memory problems if reading records from Kinesis is faster than processing them by Beam.